WalletWasabi.Fluent/Models/WebcamQrReader.cs

```diff
@@ -0,0 +1,166 @@
+using System.Threading;
+
+namespace WalletWasabi.Fluent.Models
+{
+	public class WebcamQrReader
+	{
+		public bool RequestEnd { get; set; }
+		public Network Network { get; }
+		public Task? ScanningTask { get; set; }
+
+		public WebcamQrReader(Network network)
+		{
+			Network = network;
+		}
+
+		public void StartScanning()
+		{
+			if (ScanningTask is not { })
+			{
+				ScanningTask = Task.Run(() =>
+				{
+					VideoCapture? camera = null;
+					try
+					{
+						if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+						{
+							throw new NotImplementedException("This operating system is not supported.");
+						}
+						camera = OpenCamera();
+						RequestEnd = false;
+						Scan(camera);
+					}
+					catch (Exception exc)
+					{
+						Logger.LogError("QR scanning stopped. Reason:", exc);
+						ErrorOccured?.Invoke(this, exc);
+					}
+					finally
+					{
+						camera?.Release();
+					}
+				});
+			}
+		}
+
+		public async Task StopScanningAsync()
+		{
+			if (ScanningTask is { } task)
+			{
+				RequestEnd = true;
+				await task;
+
+				ScanningTask = null;
+			}
+		}
+
+		private void Scan(VideoCapture camera)
+		{
+			WriteableBitmap? lastBitmap = null;
+			WriteableBitmap? currentBitmap = null;
+			using QRCodeDetector qRCodeDetector = new();
+			while (!RequestEnd)
+			{
+				try
+				{
+					using Mat frame = new();
+					camera.Read(frame);
+					if (frame.Empty() || frame.Width == 0 || frame.Height == 0)
+					{
+						continue;
+					}
+					currentBitmap = ConvertMatToWriteableBitmap(frame);
+
+					NewImageArrived?.Invoke(this, currentBitmap);
+					lastBitmap?.Dispose();
+					lastBitmap = currentBitmap;
+
+					if (qRCodeDetector.Detect(frame, out Point2f[] points))
+					{
+						string qrCode = qRCodeDetector.Decode(frame, points, new Mat());
+						if (string.IsNullOrWhiteSpace(qrCode))
+						{
+							continue;
+						}
+						if (AddressStringParser.TryParse(qrCode, Network, out _))
+						{
+							BitcoinAddressFound?.Invoke(this, qrCode);
+							break;
+						}
+						else
+						{
+							InvalidAddressFound?.Invoke(this, qrCode);
+						}
+					}
+				}
+				catch (OpenCVException exc)
+				{
+					Logger.LogWarning(exc);
+					currentBitmap?.Dispose();
+				}
+			}
+			lastBitmap?.Dispose();
+			currentBitmap?.Dispose();
+		}
+
+		private VideoCapture OpenCamera()
+		{
+			VideoCapture camera = new();
+			if (!camera.Open(0))
+			{
+				throw new InvalidOperationException("Could not open webcam.");
+			}
+			return camera;
+		}
+
+		private WriteableBitmap ConvertMatToWriteableBitmap(Mat frame)
+		{
+			PixelSize pixelSize = new(frame.Width, frame.Height);
+			Vector dpi = new(96, 96);
+			Avalonia.Platform.PixelFormat pixelFormat = Avalonia.Platform.PixelFormat.Rgba8888;
+			Avalonia.Platform.AlphaFormat alphaFormat = Avalonia.Platform.AlphaFormat.Unpremul;
+			var writeableBitmap = new WriteableBitmap(pixelSize, dpi, pixelFormat, alphaFormat);
+
+			using (var fb = writeableBitmap.Lock())
```

Comment:
I wonder what is the purpose of this lock call.

Reply:
We need to call the `Lock()` function to access the correct parameters and the `Address` that we need to copy the new image.

Reply:
Given https://github.com/zkSNACKs/WalletWasabi/pull/6002/files#r670446317, I believe this method should accept `WriteableBitmap` as a parameter.

Reply:
> We need to call the `Lock()` function to access the correct parameters and the `Address` that we need to copy the new image.

I see. That's correct.